src/datadog/grok/filters/keyvalue.rs

```diff
@@ -155,145 +196,70 @@ pub fn apply_filter(value: &Value, filter: &KeyValueFilter) -> Result<Value, Gro
     }
 }
 
-type SResult<'a, O> = IResult<&'a str, O, (&'a str, nom::error::ErrorKind)>;
-
-#[inline]
-fn parse<'a>(
-    input: &'a str,
-    key_value_delimiter: &'a str,
-    field_delimiters: &'a [&'a str],
-    quotes: &'a [(char, char)],
-    value_re: &Regex,
-) -> Result<Vec<(String, Value)>, String> {
-    let (rest, result) = parse_line(
-        input,
-        key_value_delimiter,
-        field_delimiters,
-        quotes,
-        value_re,
-    )
-    .map_err(|_| format!("could not parse '{}' as 'keyvalue'", input))?;
-
-    if rest.trim().is_empty() {
-        Ok(result)
-    } else {
-        Err("could not parse whole line successfully".into())
-    }
-}
-
-/// Parse the line as a separated list of key value pairs.
-#[inline]
-fn parse_line<'a>(
-    input: &'a str,
-    key_value_delimiter: &'a str,
-    field_delimiters: &'a [&'a str],
-    quotes: &'a [(char, char)],
-    value_re: &'a Regex,
-) -> SResult<'a, Vec<(String, Value)>> {
-    let mut last_result = None;
-    for &field_delimiter in field_delimiters {
-        match separated_list1(
-            parse_field_delimiter(field_delimiter),
-            parse_key_value(key_value_delimiter, field_delimiter, quotes, value_re),
-        )(input)
-        {
-            Ok((rest, v)) if rest.trim().is_empty() => {
-                return Ok((rest, v));
+fn parse_key_value_capture(
+    filter: &KeyValueFilter,
+    result: &mut Value,
+    c: Result<Captures, fancy_regex::Error>,
+) {
+    let key = parse_key(extract_capture(&c, 1), filter.quotes.as_slice());
+    if !key.contains(' ') {
+        let value = extract_capture(&c, 2);
+        // trim trailing comma for value
+        let value = value.trim_end_matches(|c| c == ',');
+
+        if let Ok((_, value)) = parse_value(value, filter.quotes.as_slice()) {
+            if !(value.is_null()
+                || matches!(&value, Value::Bytes(b) if b.is_empty())
+                || key.is_empty())
+            {
+                let path = crate::path!(key);
+                match result.get(path).cloned() {
+                    Some(Value::Array(mut values)) => {
+                        values.push(value);
+                        result.insert(path, values);
+                    }
+                    Some(prev) => {
+                        result.insert(path, Value::Array(vec![prev, value]));
+                    }
+                    None => {
+                        result.insert(path, value);
+                    }
+                };
```

Comment:
The cloning here turns into some nasty O(N^2) behavior for array values, not to mention reinserting the same value at an existing slot. Could you use the `entry` interface to avoid needing to clone and reinsert?